pkg/orchestrator/etcd_worker.go

```diff
@@ -128,28 +142,28 @@ func (worker *EtcdWorker) Run(ctx context.Context, timerInterval time.Duration)
 			// `applyPatches` is all-or-none, so in case of success, we should clear all the pendingPatches.
 			pendingPatches = pendingPatches[:0]
 		} else {
+			if exiting {
+				// If exiting is true here, it means that the reactor returned `ErrReactorFinished` last tick, and all pending patches is applied.
+				return nil
+			}
 			if worker.revision < worker.barrierRev {
 				// We hold off notifying the Reactor because barrierRev has not been reached.
 				// This usually happens when a committed write Txn has not been received by Watch.
 				continue
 			}
 
 			// We are safe to update the ReactorState only if there is no pending patch.
-			for _, update := range worker.pendingUpdates {
-				err := worker.state.Update(update.key, update.value, false)
-				if err != nil {
-					return errors.Trace(err)
-				}
+			if err := worker.applyUpdates(); err != nil {
+				return errors.Trace(err)
 			}
 
-			worker.pendingUpdates = worker.pendingUpdates[:0]
 			nextState, err := worker.reactor.Tick(ctx, worker.state)
 			if err != nil {
-				if errors.Cause(err) == cerrors.ErrReactorFinished {
-					// normal exit
-					return nil
+				if errors.Cause(err) != cerrors.ErrReactorFinished {
```

Comment:
error used here does not follow the best practice, since `cerrors.ErrReactorFinished` is a global error generator, we should generate an error via `GenWithStackByArgs` or `GenWithStack` and check error by `cerrors.ErrReactorFinished.Equal(err)`